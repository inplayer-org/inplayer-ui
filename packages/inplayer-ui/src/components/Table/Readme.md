Table is a component for displaying information in order user to look for patterns and easy to scan details.

<br /> The columns contain `title` - which is presented at the top od the columns and `key` for each column.
<br /> The data props contain object of each keys that are presented in the table cell according to the key.
<br /> The `options` object contains table options like row selection and row actions.

```jsx static
import React from 'react';
import { Table } from '@inplayer-org/inplayer-ui';

class Page extends React.Component {
  constructor() {
    super();

    this.state = {
      page: 0,
    };

    this.onPageChange = this.onPageChange.bind(this);
  }

  onPageChange(data) {
    this.setState({ page: data.selected });
  }

  render() {
    const columns = [
      {
        title: 'ID',
        key: 'id',
      },
      {
        title: '',
        key: 'icon',
        render: ({ value }) => <Icon name={value} />,
      },
      {
        title: 'Name',
        key: 'name',
      },
      {
        title: 'Date Created',
        key: 'dateCreated',
      },
      {
        title: 'Director',
        key: 'director',
        render: ({ value, rowValues }) => `${value} - ${rowValues.dateCreated}`,
      },
    ];

    const data = [
      {
        id: 22,
        icon: 'music',
        name: "Won't You Be My Neighbour?",
        director: 'Morgan Neville',
        dateCreated: '2018/07/24',
      },
      {
        id: 28,
        icon: 'code',
        name: 'Soundtrack: No Country',
        director: 'Coen Brothers',
        dateCreated: '2018/07/21',
      },
    ];

    const totalPages = Math.ceil(data.length / 5);

    return (
      <React.Fragment>
        <Table
          columns={columns}
          data={data}
          options={{
            rowSelection: {
              active: true,
              action: data => console.log(data),
            },
            rowActions: [
              { icon: 'edit', onClick: id => console.log(id) },
              { icon: 'cog', onClick: id => console.log(id) },
              { icon: 'trash', onClick: id => console.log(id) },
            ],
          }}
        />
        <Pagination
          pageCount={totalPages}
          currentPage={this.state.page}
          pageRangeDisplayed={5}
          marginPagesDisplayed={5}
          previousLabel="PREV"
          nextLabel="NEXT"
          breakLabel="..."
          onPageChange={this.onPageChange}
        />
      </React.Fragment>
    );
  }
}
```

```jsx
const columns = [
  {
    title: 'ID',
    key: 'id',
  },
  {
    title: '',
    key: 'icon',
    render: ({ value }) => <Icon name={value} />,
  },
  {
    title: 'Name',
    key: 'name',
  },
  {
    title: 'Date Created',
    key: 'dateCreated',
  },
  {
    title: 'Director',
    key: 'director',
    render: ({ value, rowValues }) => `${value} - ${rowValues.dateCreated}`,
  },
];

const data = [
  {
    id: 22,
    icon: 'music',
    name: "Won't You Be My Neighbour?",
    director: 'Morgan Neville',
    dateCreated: '2018/07/24',
  },
  {
    id: 28,
    icon: 'code',
    name: 'Soundtrack: No Country',
    director: 'Coen Brothers',
    dateCreated: '2018/07/21',
  },
];

const totalPages = Math.ceil(data.length / 5);

class TableExample extends React.Component {
  constructor() {
    super();

    this.state = {
      page: 0,
    };

    this.onPageChange = this.onPageChange.bind(this);
  }

  onPageChange(data) {
    this.setState({ page: data.selected });
  }

  render() {
    return (
      <React.Fragment>
        <Typography variant="h2">With array of row actions</Typography>
        <Input
          type="search"
          placeholder="Search"
          onChange={() => console.log('onChange')}
          style={{ maxWidth: 300 }}
        />
        <br />
        <Table
          columns={columns}
          data={data}
          options={{
            rowSelection: {
              active: true,
              action: data => console.log(data),
            },
            rowActions: [
              { icon: 'edit', onClick: id => console.log(id) },
              { icon: 'cog', onClick: id => console.log(id) },
              { icon: 'trash', onClick: id => console.log(id) },
            ],
          }}
        />
        <Pagination
          pageCount={totalPages}
          currentPage={this.state.page}
          pageRangeDisplayed={5}
          marginPagesDisplayed={5}
          previousLabel="PREV"
          nextLabel="NEXT"
          breakLabel="..."
          onPageChange={this.onPageChange}
        />
      </React.Fragment>
    );
  }
}
<TableExample />;
```

```jsx
const columns = [
  {
    title: 'ID',
    key: 'id',
  },
  {
    title: '',
    key: 'icon',
    render: ({ value }) => <Icon name={value} />,
  },
  {
    title: 'Name',
    key: 'name',
  },
  {
    title: 'Date Created',
    key: 'dateCreated',
  },
  {
    title: 'Director',
    key: 'director',
    render: ({ value, rowValues }) => `${value} - ${rowValues.dateCreated}`,
  },
];

const data = [
  {
    id: 22,
    icon: 'music',
    name: "Won't You Be My Neighbour?",
    director: 'Morgan Neville',
    dateCreated: '2018/07/24',
  },
  {
    id: 28,
    icon: 'code',
    name: 'Soundtrack: No Country',
    director: 'Coen Brothers',
    dateCreated: '2018/07/21',
  },
];

const totalPages = Math.ceil(data.length / 5);

class TableExample extends React.Component {
  constructor() {
    super();

    this.state = {
      page: 0,
    };

    this.onPageChange = this.onPageChange.bind(this);
  }

  onPageChange(data) {
    this.setState({ page: data.selected });
  }

  render() {
    return (
      <React.Fragment>
        <Typography variant="h2">With rowActions as render prop</Typography>
        <Input
          type="search"
          placeholder="Search"
          onChange={() => console.log('onChange')}
          style={{ maxWidth: 300 }}
        />
        <br />
        <Table
          columns={columns}
          data={data}
          options={{
            rowSelection: {
              active: true,
              action: data => console.log(data),
            },
            rowActions: ({ row }) => <Icon name={row.icon} />,
          }}
          tableButton={{
            label: 'Add new',
            icon: <InPlayerIcon name="add" />,
            onClick: () => console.log('Clicked'),
          }}
        />
        <Pagination
          pageCount={totalPages}
          currentPage={this.state.page}
          pageRangeDisplayed={5}
          marginPagesDisplayed={5}
          previousLabel="PREV"
          nextLabel="NEXT"
          breakLabel="..."
          onPageChange={this.onPageChange}
        />
      </React.Fragment>
    );
  }
}
<TableExample />;
```

```jsx
const columns = [
  {
    title: 'ID',
    key: 'id',
  },
  {
    title: '',
    key: 'icon',
    render: ({ value }) => <Icon name={value} />,
  },
  {
    title: 'Name',
    key: 'name',
  },
  {
    title: 'Date Created',
    key: 'dateCreated',
  },
  {
    title: 'Director',
    key: 'director',
    render: ({ value, rowValues }) => `${value} - ${rowValues.dateCreated}`,
  },
];

const data = [
  {
    id: 22,
    icon: 'music',
    name: "Won't You Be My Neighbour?",
    director: 'Morgan Neville',
    dateCreated: '2018/07/24',
  },
  {
    id: 28,
    icon: 'code',
    name: 'Soundtrack: No Country',
    director: 'Coen Brothers',
    dateCreated: '2018/07/21',
  },
];

const totalPages = Math.ceil(data.length / 5);

class TableExample extends React.Component {
  constructor() {
    super();

    this.state = {
      page: 0,
    };

    this.onPageChange = this.onPageChange.bind(this);
  }

  onPageChange(data) {
    this.setState({ page: data.selected });
  }

  renderHeaderSection() {
    const options = [
      { value: 'Option 1', displayName: 'Option 1' },
      { value: 'Option 2', displayName: 'Option 2' },
    ];

    return (
      <div>
        <Dropdown
          options={options}
          value={options[0].value}
          onChange={console.log}
        />
        <Dropdown
          options={options}
          value={options[0].value}
          onChange={console.log}
        />
      </div>
    );
  }

  render() {
    return (
      <React.Fragment>
        <Typography variant="h2">With Header Section</Typography>
        <br />
        <Table
          columns={columns}
          data={data}
          options={{
            rowSelection: {
              active: true,
              action: data => console.log(data),
            },
            rowActions: ({ row }) => <Icon name={row.icon} />,
            headerSection: this.renderHeaderSection(),
          }}
          tableButton={{
            label: 'Add new',
            icon: <InPlayerIcon name="add" />,
            onClick: () => console.log('Clicked'),
          }}
        />
        <Pagination
          pageCount={totalPages}
          currentPage={this.state.page}
          pageRangeDisplayed={5}
          marginPagesDisplayed={5}
          previousLabel="PREV"
          nextLabel="NEXT"
          breakLabel="..."
          onPageChange={this.onPageChange}
        />
      </React.Fragment>
    );
  }
}
<TableExample />;
```
