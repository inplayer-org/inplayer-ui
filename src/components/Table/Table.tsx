import 'react-dates/initialize';
import React, { ReactNode, SyntheticEvent, Component } from 'react';
import { CSSProperties } from 'styled-components';
import Grid from '../Grid';
import Loader from '../Loader';
import {
  ButtonTableRow,
  LoaderContainer,
  TableButton,
  TableCell,
  TableRow,
  TableCheckbox,
  TableHeadRow,
  TableHeaderCell,
  TableWithHeaderSectionContainer,
  TableWrapper,
} from './styled';

interface Data {
  id: string;
  [row: string]: any;
}

type Render = {
  value: string;
  rowValues: Data;
};

type Column = {
  title: string;
  key: string;
  render: ({ value, rowValues }: Render) => ReactNode;
  style?: CSSProperties;
  alignRight?: any;
};

type RowActions<T> =
  | Array<{
      icon: string | Node;
      onClick: (id: number | string) => any;
      render?: any;
    }>
  | ((props: { row: T }) => Node);

type TableOptions<T> = {
  rowSelection: {
    active: boolean;
    action: (selectedItems: Array<T>) => any;
  };
  rowActions: RowActions<Data>;
  headerSection?: Node;
};

// eslint-disable-next-line @typescript-eslint/no-use-before-define
type Props<T = Data> = {
  columns: Array<Column> | any;
  data: Array<any>;
  className?: string;
  style?: CSSProperties;
  options: TableOptions<T>;
  showLoader?: boolean;
  renderEmptyTable?: boolean;
  tableButton?: {
    label: string;
    icon?: string | Node;
    onClick: (e: SyntheticEvent) => any;
    type: string;
  };
};

type State = {
  selected: {
    [key in number | string]: any;
  };
  selectedAll: boolean;
};

const rowActionsExist = (actions: RowActions<any>) =>
  typeof actions === 'function' ||
  (typeof actions === 'object' && Array.isArray(actions) && actions.length);

class Table<T> extends Component<Props<T>, State> {
  // eslint-disable-next-line react/static-property-placement
  static defaultProps = {
    className: '',
    style: {},
    options: {
      rowSelection: {
        active: false,
      },
      rowActions: [],
      headerSection: undefined,
    },
    showLoader: false,
    renderEmptyTable: false,
    tableButton: null,
  };

  state: State = {
    selected: {},
    selectedAll: false,
  };

  toggleRow = (id: number | string) => () => {
    const { selected } = this.state;
    const isSelected = !selected[id];
    const newSelected = {
      ...selected,
      [id]: isSelected,
    };

    this.setState(({ selectedAll }) => ({
      selected: newSelected,
      selectedAll: !isSelected ? false : selectedAll,
    }));
  };

  toggleSelectAll = () => {
    const { selectedAll } = this.state;
    const { data } = this.props;
    const selected = data.reduce((acc, current) => ({ ...acc, [current.id]: !selectedAll }), {});

    this.setState(({ selectedAll: prevSelectedAll }) => ({
      selected,
      selectedAll: !prevSelectedAll,
    }));
  };

  generateRows = (data: Array<Data>) => {
    const { options } = this.props;
    const { rowSelection, rowActions }: TableOptions<T> = options;
    let newData = [...data];

    if (rowSelection && rowSelection.active) {
      const { selected } = this.state;

      newData = data.map((dataCell): any => ({
        ...dataCell,
        check: (
          <TableCheckbox
            id={dataCell.id}
            checked={selected[dataCell.id]}
            onChange={this.toggleRow(dataCell.id)}
          />
        ),
      }));
    }

    if (rowActionsExist(rowActions)) {
      newData = newData.map((dataCell) => {
        const actionsContent =
          typeof rowActions === 'function'
            ? rowActions({ row: dataCell })
            : rowActions.map((action) => {
                if (!action.render) return;

                return action.render({ row: dataCell });
              });

        return {
          ...dataCell,
          actions: actionsContent,
        };
      });
    }

    return newData;
  };

  generateColumns = (columns: Array<Column>) => {
    const { selectedAll } = this.state;
    const { options } = this.props;
    const { rowSelection, rowActions } = options;

    let newColumns: any = [...columns];

    if (rowSelection && rowSelection.active) {
      const rowCheckbox: any = {
        title: (
          <TableCheckbox id="toggleAll" checked={selectedAll} onChange={this.toggleSelectAll} />
        ),
        key: 'check',
      };

      newColumns = [rowCheckbox, ...columns];
    }

    if (rowActionsExist(rowActions)) {
      const actionsColumn = {
        title: 'Actions',
        key: 'actions',
        alignRight: true,
      };
      newColumns = [...newColumns, actionsColumn];
    }

    return newColumns;
  };

  renderColumns = (columns: Array<Column>): Array<JSX.Element> =>
    this.generateColumns(columns).map((column: Column, index: number) => (
      <TableHeaderCell key={index} alignRight={column.alignRight}>
        {column.title}
      </TableHeaderCell>
    ));

  renderData = (columns: Array<Column>, data: Array<Data>) => {
    const newColumns = this.generateColumns(columns);
    const newData = this.generateRows(data);

    const { tableButton } = this.props;

    return newData.map((row) => (
      <TableRow key={row.id} noBottomBorder={!tableButton}>
        {newColumns.map((column: Column, index: number) => (
          <TableCell
            key={index}
            isActionsCell={column.key === 'actions'}
            style={column.style || {}}
          >
            {column.render
              ? column.render({ value: row[column.key], rowValues: row })
              : row[column.key]}
          </TableCell>
        ))}
      </TableRow>
    ));
  };

  renderTable = () => {
    const {
      columns,
      data,
      className,
      style,
      showLoader,
      renderEmptyTable,
      tableButton,
      options: { headerSection },
    } = this.props;

    if (showLoader) {
      return (
        <LoaderContainer>
          <Loader />
        </LoaderContainer>
      );
    }

    if (!data.length && !renderEmptyTable) {
      return null;
    }

    const columnContent = this.renderColumns(columns);

    const hasHeaderSection = typeof headerSection !== 'undefined';

    return (
      <TableWrapper className={className} style={style} hasHeaderSection={hasHeaderSection}>
        <thead>
          <TableHeadRow>{columnContent}</TableHeadRow>
        </thead>
        <tbody>{this.renderData(columns, data)}</tbody>
        <tfoot>
          {tableButton && (
            <ButtonTableRow>
              <TableCell colSpan={columnContent.length}>
                <TableButton
                  modifiers={['buttonLink']}
                  fullWidth
                  fullHeight
                  onClick={tableButton.onClick}
                  icon={tableButton.icon}
                  iconPosition="left"
                  type={tableButton.type}
                >
                  {tableButton.label}
                </TableButton>
              </TableCell>
            </ButtonTableRow>
          )}
        </tfoot>
      </TableWrapper>
    );
  };

  renderTableWithHeaderSection = () => {
    const {
      options: { headerSection },
    } = this.props;

    return (
      <TableWithHeaderSectionContainer columns={1}>
        <Grid.Cell>{headerSection}</Grid.Cell>
        <Grid.Cell>{this.renderTable()}</Grid.Cell>
      </TableWithHeaderSectionContainer>
    );
  };

  render() {
    const {
      options: { headerSection },
    } = this.props;

    if (!headerSection) return this.renderTable();

    return this.renderTableWithHeaderSection();
  }
}

export default Table;
