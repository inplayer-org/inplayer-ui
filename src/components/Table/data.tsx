import React from 'react';
import { FaMusic, FaCode } from 'react-icons/fa';

export const columns = [
  {
    title: 'ID',
    key: 'id',
  },
  {
    title: '',
    key: 'icon',
    render: () => <FaMusic />,
  },
  {
    title: 'Name',
    key: 'name',
  },
  {
    title: 'Date Created',
    key: 'dateCreated',
  },
  {
    title: 'Director',
    key: 'director',
    render: ({ value, rowValues }: any) => `${value} - ${rowValues.dateCreated}`,
  },
];

export const data = [
  {
    id: 2,
    icon: <FaMusic />,
    name: "Won't You Be My Neighbour?",
    director: 'Morgan Neville',
    dateCreated: '2018/07/24',
  },
  {
    id: 8,
    icon: <FaCode />,
    name: 'Soundtrack: No Country',
    director: 'Coen Brothers',
    dateCreated: '2018/07/21',
  },
];

export const totalPages = Math.ceil(data.length / 5);
