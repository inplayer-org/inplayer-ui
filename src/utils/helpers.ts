import moment from 'moment';

export const getMonthOptions = (): Array<any> =>
  moment.months().map((label, value) => ({ value, displayName: label }));

export const getYearOptions = () => {
  const yearOptions: Array<any> = [];
  for (let i = moment().year() - 100; i <= moment().year(); i += 1) {
    yearOptions.push({ value: i.toString(), displayName: i.toString() });
  }

  return yearOptions;
};
