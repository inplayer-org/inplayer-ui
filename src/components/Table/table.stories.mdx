import { useState } from 'react';
import { Meta, Story, Preview, Props } from "@storybook/addon-docs/blocks";
import Table from './'
import { MdDelete, MdErrorOutline } from 'react-icons/md';
import Pagination from '../Pagination'
import { columns, data } from './data.tsx'

<Meta title="Table" component={<Table/>} />


```jsx
import { Table } from '@inplayer-org/inplayer-ui';
const Page = () => (
  <Table />
)
```

<Preview withSource="open">
  <Story name="default">
    <>
    <Table
            columns={columns}
            data={data}
            options={{
                rowSelection: {
                    active: true,
                    action: (data) => console.log(data),
                },
                rowActions: [
                    { icon: <MdDelete />, onClick: (id) => console.log(id), render: () => (<p>render</p>)},
                      { icon: "cog", onClick: (id) => console.log(id), render: () => <p>render</p>},
                      { icon: "trash", onClick: (id) => console.log(id) },
                ],
            }}
        />
        <Pagination
            onPageChange={console.log}
            totalItems={50}
            startPage={4}
            numberOfPagesDisplayed={10}
            itemsPerPage={12}
        />
      </>
  </Story>
</Preview>