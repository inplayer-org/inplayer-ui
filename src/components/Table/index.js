// @flow
import React, { type Node } from 'react';
import styled, { css } from 'styled-components';
import { ifProp } from 'styled-tools';
import colors from 'config/colors';
import { uiColors, fontSizes, fontWeights } from 'utils';
import Checkbox from 'components/Checkbox';
import Icon from 'elements/Icon';
import Loader from 'elements/Loader/index';
import Button from 'elements/Button/index';

const TableWrapper = styled.table`
  font-family: Roboto, sans-serif;
  background: ${colors.white};
  width: 100%;
  text-align: left;
  color: ${uiColors('text.light')};
  box-sizing: border-box;
  border: 1px solid ${colors.gray};
  border-radius: 3px;
  position: relative;
  font-weight: ${fontWeights('light')};
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  border-collapse: collapse;
  table-layout: auto;
`;

const TableHeadRow = styled.tr`
  border-bottom: 1px solid ${colors.gray};
  width: 100%;
`;

const TableRow = styled.tr`
  border-bottom: 1px solid ${colors.lightGray};
  width: 100%;
  ${({ noBottomBorder }) =>
    noBottomBorder &&
    css`
      &:last-child {
        border-bottom: none;
      }
    `}
`;

const ButtonTableRow = styled(TableRow)`
  background: ${colors.lightGray};
  width: 100%;
  border-bottom: 1px solid ${colors.gray};
  border-top: 1px solid ${colors.gray};
`;

const TableCell = styled.td`
  font-size: ${fontSizes('medium')};
  padding: 0.4375rem 1%;
  vertical-align: middle;
  line-height: 30px;
  text-align: ${ifProp('isActionsCell', 'right', 'left')};
  font-weight: ${fontWeights('light')};
  word-wrap: break-word;
  text-overflow: ellipsis;
  white-space: nowrap;

  a {
    color: ${colors.fontGray};

    &:hover {
      color: ${colors.fontDarkGray};
    }
  }
`;

const TableHeaderCell = styled.th`
  padding: 1rem 1%;
  font-weight: ${fontWeights('light')};
  font-size: ${fontSizes('small')};
  text-align: ${({ alignRight }) => (alignRight ? 'right' : 'left')};
`;

const ActionIcon = styled(Icon)`
  padding: 0.125rem;
  cursor: pointer;
  transition: all 0.3s ease;

  &:hover {
    color: ${uiColors('primary.main')};
  }
`;

const LoaderContainer = styled.div`
  width: 100%;
  min-height: 300px;
  display: flex;
  justify-content: center;
  align-items: center;
`;

const TableCheckbox = styled(Checkbox)`
  label {
    top: -10px;
  }
`;

const TableButton = styled(Button)`
  display: flex;
  align-items: center;
  justify-content: flex-start;
  color: ${colors.fontGray};

  &:hover {
    color: ${({ theme }) => theme.palette.primary.main};
  }
`;

interface Data {
  id: string;
}

type Column = {
  title: string,
  key: string,
  render: ({ value: string, rowValues: Data }) => Node,
  style?: Object,
};

type RowActions<T> =
  | Array<{
      icon: string,
      onClick: (id: number | string) => any,
    }>
  | ((props: { row: T }) => Node);

type TableOptions<T> = {
  rowSelection: {
    active: boolean,
    action: (selectedItems: Array<T>) => any,
  },
  rowActions: RowActions<T>,
};

type Props<T = Data> = {
  columns: Array<Column>,
  data: Array<T>,
  className?: string,
  style?: Object,
  options?: TableOptions<T>,
  showLoader?: boolean,
  tableButton?: {
    label: string,
    icon?: string | Node,
    onClick: (e: SyntheticEvent) => any,
    type: string,
  },
};

type State = {
  selected: {
    [number | string]: boolean,
  },
  selectedAll: boolean,
};

const rowActionsExist = (actions: RowActions) =>
  typeof actions === 'function' ||
  (typeof actions === 'object' && Array.isArray(actions) && actions.length);

class Table<T> extends React.Component<Props<T>, State> {
  state = {
    selected: {},
    selectedAll: false,
  };

  toggleRow = (id: number) => () => {
    const { selected } = this.state;
    const isSelected = !selected[id];
    const newSelected = {
      ...selected,
      [id]: isSelected,
    };

    this.setState(({ selectedAll }) => ({
      selected: newSelected,
      selectedAll: !isSelected ? false : selectedAll,
    }));
  };

  toggleSelectAll = () => {
    const { selectedAll } = this.state;
    const { data } = this.props;
    const selected = data.reduce((acc, current) => ({ ...acc, [current.id]: !selectedAll }), {});

    this.setState(({ selectedAll: prevSelectedAll }) => ({
      selected,
      selectedAll: !prevSelectedAll,
    }));
  };

  generateRows = (data: Array<Data>) => {
    const { options } = this.props;
    const { rowSelection, rowActions } = options;
    let newData = [...data];

    if (rowSelection && rowSelection.active) {
      const { selected } = this.state;

      newData = data.map(dataCell => ({
        ...dataCell,
        check: (
          <TableCheckbox
            id={dataCell.id}
            checked={selected[dataCell.id]}
            onChange={this.toggleRow(dataCell.id)}
          />
        ),
      }));
    }

    if (rowActionsExist(rowActions)) {
      newData = newData.map(dataCell => {
        const actionsContent =
          typeof rowActions === 'function'
            ? rowActions({ row: dataCell })
            : rowActions.map((action, index) => {
                if (action.render) {
                  return action.render({ row: dataCell });
                }
                return (
                  <ActionIcon
                    key={index}
                    name={action.icon}
                    onClick={() => action.onClick(dataCell.id)}
                  />
                );
              });
        return {
          ...dataCell,
          actions: actionsContent,
        };
      });
    }

    return newData;
  };

  generateColumns = (columns: Array<Column>) => {
    const { selectedAll } = this.state;
    const { options } = this.props;
    const { rowSelection, rowActions } = options;

    let newColumns = [...columns];

    if (rowSelection && rowSelection.active) {
      const rowCheckbox = {
        title: (
          <TableCheckbox id="toggleAll" checked={selectedAll} onChange={this.toggleSelectAll} />
        ),
        key: 'check',
      };

      newColumns = [rowCheckbox, ...columns];
    }

    if (rowActionsExist(rowActions)) {
      const actionsColumn = { title: 'Actions', key: 'actions', alignRight: true };
      newColumns = [...newColumns, actionsColumn];
    }

    return newColumns;
  };

  renderColumns = (columns: Array<Column>): Array<Node> =>
    this.generateColumns(columns).map((column, index) => (
      <TableHeaderCell key={index} alignRight={column.alignRight}>
        {column.title}
      </TableHeaderCell>
    ));

  renderData = (columns: Array<Column>, data: Array<Data>) => {
    const newColumns = this.generateColumns(columns);
    const newData = this.generateRows(data);

    const { tableButton } = this.props;

    return newData.map(row => (
      <TableRow key={row.id} noBottomBorder={!tableButton}>
        {newColumns.map((column, index) => (
          <TableCell
            key={index}
            isActionsCell={column.key === 'actions'}
            style={column.style || {}}
          >
            {column.render
              ? column.render({ value: row[column.key], rowValues: row })
              : row[column.key]}
          </TableCell>
        ))}
      </TableRow>
    ));
  };

  render() {
    const { columns, data, className, style, showLoader, tableButton } = this.props;

    const columnContent = this.renderColumns(columns);

    return (
      <TableWrapper className={className} style={style}>
        {showLoader ? (
          <LoaderContainer>
            <Loader />
          </LoaderContainer>
        ) : (
          <>
            <thead>
              <TableHeadRow>{columnContent}</TableHeadRow>
            </thead>
            <tbody>{this.renderData(columns, data)}</tbody>
            <tfoot>
              {tableButton && (
                <ButtonTableRow>
                  <TableCell colSpan={columnContent.length}>
                    <TableButton
                      modifiers={['buttonLink']}
                      fullWidth
                      fullHeight
                      onClick={tableButton.onClick}
                      icon={tableButton.icon}
                      iconPosition="left"
                      type={tableButton.type}
                    >
                      {tableButton.label}
                    </TableButton>
                  </TableCell>
                </ButtonTableRow>
              )}
            </tfoot>
          </>
        )}
      </TableWrapper>
    );
  }
}

Table.defaultProps = {
  className: '',
  style: {},
  options: {
    rowSelection: {
      active: false,
    },
    rowActions: [],
  },
  showLoader: false,
  tableButton: null,
};

export default Table;
