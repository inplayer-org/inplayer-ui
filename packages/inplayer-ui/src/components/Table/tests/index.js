import { generateSnapshots } from 'packages/inplayer-ui/src/components/Table/tests/helpers';

import Table from '../index';

generateSnapshots(Table, {
  columns: [
    {
      title: 'Name',
      key: 'name',
    },
  ],
  data: [{ name: 'Eli' }],
  options: {
    rowSelection: {
      active: true,
      action: data => console.log(data),
    },
    rowActions: [
      { icon: 'edit', onClick: id => console.log(id) },
      { icon: 'cog', onClick: id => console.log(id) },
      { icon: 'trash', onClick: id => console.log(id) },
    ],
  },
});
