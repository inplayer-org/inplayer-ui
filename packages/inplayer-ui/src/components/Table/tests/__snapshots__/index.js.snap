// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table renders correctly 1`] = `
.c5 {
  -webkit-text-decoration: none;
  text-decoration: none;
  font-weight: 300;
  cursor: pointer;
  vertical-align: middle;
  color: #3e4348;
  -webkit-transition: all ease 300ms;
  transition: all ease 300ms;
  line-height: 20px;
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.c5:not(:disabled):hover {
  color: #7d7e7f;
}

.c4 > input {
  display: none;
}

.c4 > input + label {
  position: relative;
  padding-left: 1.5rem;
  cursor: pointer;
}

.c4 > input + label::before {
  content: '';
  position: absolute;
  left: 0;
  width: 1rem;
  height: 1rem;
  border: 1px solid #d7dde5;
  background: #fff;
  border-radius: 2px;
  -webkit-transition: all ease 300ms;
  transition: all ease 300ms;
}

.c4 > input + label::after {
  content: '';
  position: absolute;
  top: 5px;
  left: 4px;
  -webkit-transition: all ease 300ms;
  transition: all ease 300ms;
}

.c4 > input:checked + label::before {
  border: 1px solid #07AAE6;
}

.c4 > input:checked + label::after {
  border: 2px solid #07AAE6;
  border-top: none;
  border-right: none;
  width: 0.5rem;
  height: 0.25rem;
  opacity: 1;
  -webkit-transform: scale(1) rotate(-45deg);
  -ms-transform: scale(1) rotate(-45deg);
  transform: scale(1) rotate(-45deg);
}

.c4 > input:not(:checked) + label::after {
  border: 2px solid #fff;
  border-top: none;
  border-right: none;
  width: 0.5rem;
  height: 0.25rem;
  opacity: 1;
  -webkit-transform: scale(0);
  -ms-transform: scale(0);
  transform: scale(0);
}

.c4 > input:checked + label:hover::before {
  border-color: transparent;
  background: #97CCF6;
}

.c4 > input:checked + label:hover::after {
  border-color: #07AAE6;
}

.c4 > input:not(:checked) + label:hover::before {
  border: 1px solid #07AAE6;
  background-color: #97CCF6;
}

.c4 > input:not(:checked) + label:hover::after {
  -webkit-transform: scale(1) rotate(-45deg);
  -ms-transform: scale(1) rotate(-45deg);
  transform: scale(1) rotate(-45deg);
}

.c4 > input:checked:focus + label::before,
.c4 input:not(:checked):focus + label::before {
  border: 1px dotted #97CCF6;
}

.c4 > input:disabled:checked + label {
  color: #c8ccd4;
}

.c4 > input:disabled:checked + label::before {
  border: 1px solid #d7dde5;
  background: transparent;
}

.c4 > input:disabled:checked + label::after {
  border-color: #d7dde5;
  background: transparent;
}

.c4 > input:disabled:not(:checked) + label {
  color: #c8ccd4;
}

.c4 > input:disabled:not(:checked) + label::before {
  border: 1px solid #d7dde5;
  background: transparent;
}

.c0 {
  font-family: Roboto,sans-serif;
  background: #fff;
  width: 100%;
  text-align: left;
  color: #7d7e7f;
  box-sizing: border-box;
  border: 1px solid #d7dde5;
  border-radius: 3px;
  position: relative;
  font-weight: 300;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  border-collapse: collapse;
  table-layout: auto;
}

.c1 {
  border-bottom: 1px solid #d7dde5;
  width: 100%;
}

.c7 {
  border-bottom: 1px solid #F6F8FB;
  width: 100%;
}

.c7:last-child {
  border-bottom: none;
}

.c8 {
  font-size: 1rem;
  padding: 0.4375rem 1%;
  vertical-align: middle;
  line-height: 30px;
  text-align: left;
  font-weight: 300;
  word-wrap: break-word;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.c8 a {
  color: #7d7e7f;
}

.c8 a:hover {
  color: #3e4348;
}

.c9 {
  font-size: 1rem;
  padding: 0.4375rem 1%;
  vertical-align: middle;
  line-height: 30px;
  text-align: right;
  font-weight: 300;
  word-wrap: break-word;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.c9 a {
  color: #7d7e7f;
}

.c9 a:hover {
  color: #3e4348;
}

.c2 {
  padding: 1rem 1%;
  font-weight: 300;
  font-size: 0.875rem;
  text-align: left;
}

.c6 {
  padding: 1rem 1%;
  font-weight: 300;
  font-size: 0.875rem;
  text-align: right;
}

.c10 {
  display: inline-block;
  font-size: 1rem;
  padding: 0.125rem;
  cursor: pointer;
  -webkit-transition: all 0.3s ease;
  transition: all 0.3s ease;
}

.c10:hover {
  color: #07AAE6;
}

.c3 label {
  top: -10px;
}

<table
  className="c0"
  style={Object {}}
>
  <thead>
    <tr
      className="c1"
    >
      <th
        className="c2"
      >
        <div
          className="c3 c4"
          style={Object {}}
        >
          <input
            checked={false}
            disabled={false}
            id="toggleAll"
            onChange={[Function]}
            type="checkbox"
          />
          <label
            className="c5"
            htmlFor="toggleAll"
          />
        </div>
      </th>
      <th
        className="c2"
      >
        Name
      </th>
      <th
        className="c6"
      >
        Actions
      </th>
    </tr>
  </thead>
  <tbody>
    <tr
      className="c7"
    >
      <td
        className="c8"
        style={Object {}}
      >
        <div
          className="c3 c4"
          style={Object {}}
        >
          <input
            disabled={false}
            onChange={[Function]}
            type="checkbox"
          />
          <label
            className="c5"
          />
        </div>
      </td>
      <td
        className="c8"
        style={Object {}}
      >
        Eli
      </td>
      <td
        className="c9"
        style={Object {}}
      >
        <span
          aria-hidden={true}
          className="fa fa-edit c10"
          onClick={[Function]}
        />
        <span
          aria-hidden={true}
          className="fa fa-cog c10"
          onClick={[Function]}
        />
        <span
          aria-hidden={true}
          className="fa fa-trash c10"
          onClick={[Function]}
        />
      </td>
    </tr>
  </tbody>
  <tfoot />
</table>
`;
